lib/BuildSystem/LaneBasedExecutionQueue.cpp

```diff
@@ -120,20 +150,42 @@ class LaneBasedExecutionQueue : public BuildExecutionQueue {
   
   virtual ~LaneBasedExecutionQueue() {
     // Shut down the lanes.
-    for (unsigned i = 0; i != numLanes; ++i) {
-      addJob({});
-    }
+    cancelled = true;
+    readyJobsCondition.notify_all();
+
     for (unsigned i = 0; i != numLanes; ++i) {
       lanes[i]->join();
     }
+
+    if (killAfterTimeoutThread) {
+      stopKillingCondition.notify_all();
+      killAfterTimeoutThread->join();
+    }
   }
 
   virtual void addJob(QueueJob job) override {
+    if (cancelled) {
+      // FIXME: We should eventually raise an error here as new work should not be enqueued after cancellation
+      return;
+    }
+
     std::lock_guard<std::mutex> guard(readyJobsMutex);
     readyJobs.push_back(job);
     readyJobsCondition.notify_one();
   }
 
+  virtual void cancelAllJobs() override {
+    if (cancelled) {
+      return;
+    }
+
+    cancelled = true;
```

Comment:
This can still race; what you want is:
```
auto wasAlreadyCancelled = cancelled.exchange(true);
// If we were already cancelled, do nothing.
if (wasAlreadyCancelled)
  return;

// Otherwise, initiate cancellation of the queue...
```